### ISSUE TITLE
Samba does not honor delete_old_dn=True

### ISSUE THREAD

**Issue by w-p** [cannatag/ldap3#100]
When attempting to rename an object with __connection.modify_dn()__, an error is raised with the message __ERROR, 00002035: Unwilling to perform. Old RDN must be deleted__ on Samba domain controllers.  With full debug logging on, it appears that the correct information is being transmitted to the server. This operation has been tested against Microsoft Server 2008 R2 and OpenLDAP with success.

I've had similar issues in the past with the python-ldap library but currently I cannot find the code to illustrate resolution.  This issue may be specifically Samba related but thought it should be noted here.

**Comment by w-p** [comment]
Possibly related?
http://stackoverflow.com/questions/21033254/ruby-netldap-move-user-entry-from-one-container-to-another-in-samba4

**Comment by cannatag** [comment]
Can you try with delete_old_dn=False? The rfc is quite vague on the meaning of this field. it's not clear what the meaning of 0 and 1 (the valued encoded in the message) is.

Bye,
Giovanni

**Comment by w-p** [comment]
I've attempted:
> bool: True, False
> int: 1, 0
> byte: b'1', b'0'
> hex: 0x1, 0x0

All result in:
> diagnosticMessage=b'00002035: Unwilling to perform. Old RDN must be deleted'

Again, the operation works for without issue against Active Directory, just not Samba.  On the other hand, the python-ldap library executes a rename with similar arguments without issue.  I spent a little time looking through the Samba, python-ldap, and OpenLDAP code bases but saw nothing obvious.

I'll retest later today with another MS server and try an OpenLDAP as well.

**Comment by w-p** [comment]
Just tested on my OpenLDAP server with success.

**Comment by cannatag** [comment]
Will, are you able to send me a trace of the succesful operation to openLDAP and a trace to the unsuccesful connection to Samba? You can get them with tcpdump or wireshark. The connection must be not encrypted.

If you're not able to do so I will try to set up a lab with samba, but it will take a few days for me to do that.

Bye,
Giovanni

**Comment by w-p** [comment]
I will try to do that today.  I also opened an issue with the Samba guys.
https://bugzilla.samba.org/show_bug.cgi?id=11453

**Comment by cannatag** [comment]
can you tell me your samba environment? how are you connecting to the samba server? are you using an ldap server for authentication? I would try to recreate your environement. thanks

**Comment by w-p** [comment]
Samba is the LDAP server / domain controller.  I'm using Samba 4.2.  If you are unfamiliar with it, here are some general resources.
https://wiki.samba.org/index.php/Samba_AD_DC_HOWTO

**Comment by w-p** [comment]
Here are the payloads for the modDNRequest message in python-ldap and ldap3:

python-ldap
`0/l*CN=Bill,OU=Test,DC=sh,DC=comcn=Will`

`302f0201196c2a041c434e3d42696c6c2c4f553d546573742c44433d73682c44433d636f6d0407636e3d57696c6c0101ff`

ldap3
`01l,CN=Bob,CN=Users,dc=sh,dc=com	CN=Robert`

`30310201026c2c041c434e3d426f622c434e3d55736572732c64633d73682c64633d636f6d0409434e3d526f62657274010101`

**Comment by cannatag** [comment]
Hi Will, the problem is in how samba decodes the True value of the
delete_old_dn boolean in the packet. The rules for encoding a BOOLEAN in
the ASN.1 BER encoding is as follows (from X-690 specifications): :

 Encoding of a boolean value
8.2.1
The encoding of a boolean value shall be primitive. The contents octets
shall consist of a single octet.
8.2.2
If the boolean value is:
FALSE
the octet shall be zero.

If the boolean value is
TRUE
the octet shall have any non-zero value, as a sender's option

as you can see any value different from 00 means True while 00 means False.
It seems that the samba decoder works in a different way: FF is true and
anything else is false.Probably they are mistaken by the example in the
X.690 specifications that uses the FF value, but it states that any nonzero
value.


You can get the full X.690 specification at
http://www.itu.int/ITU-T/studygroups/com17/languages/X.690-0207.pdf. The
BOOLEAN encoding is at page 6. I think you should update the bug on the
samba issue tracker whit this information.

Bye,
Giovanni



2015-08-18 22:34 GMT+02:00 Will <notifications@github.com>:

**Comment by cannatag** [comment]
Thanks, Will. It seems that the CER encoding for BOOLEAN is requested by samba (where the FF value means TRUE and the 00 value means False). LDAP protocol uses the BER encoding (where 00 means False and anything else means True). But LDAP specification states that only the FF value must be used for the True value of BOOLEAN. The encoding in ldap3 is made with the pyasn1 library, I must try to understand how to fix it. Hope to have a solution for the next release of ldap3.

Thanks a lot for your thorough debugging.

Bye,
Giovanni
.

**Comment by w-p** [comment]
Great.  Sorry that you have to change your code to deal with a special case.  I'm looking forward to the next release.

**Comment by cannatag** [comment]
The problem is that pyasn1 is not my code! Probably I have to switch to CER encoding only for some kind of ldap messages. I'll let you know.

**Comment by cannatag** [comment]
just trying a monkeypatching of pyasn1: 
Can you add the following code at the very beginning of ldap3/strategy/base.py (should be in your site-packages folder):

```python
from pyasn1.codec.ber.encoder import tagMap, BooleanEncoder
from pyasn1.type.univ import Boolean
from pyasn1.compat.octets import ints2octs


class BooleanCEREncoder(BooleanEncoder):
    _true = ints2octs((255,))

tagMap[Boolean.tagSet] = BooleanCEREncoder()
```

It should force the BooleanEncoder to emit the 0xFF value for "true" instead of 0x01.

Let me know if this works for you.

Thank,
Giovanni

**Comment by w-p** [comment]
Fantastic, works perfectly for rename and move operations.

**Comment by cannatag** [comment]
Thanks, I've refactored the code to be library-wide and pushed in dev. Hope to release it soon.Can I ask to test it again checking it out from dev?

Thanks a lot for your support.
Giovanni

**Comment by w-p** [comment]
I just tested dev against Samba and OpenLDAP successfully.  I can test MS 2012 later today but I'd guess it will pass as well.

**Comment by cannatag** [comment]
Thanks Will, I will release the fix soon.
I'm also working on a faster BER encoder/decoder to remove the dependency (or make it optional) for the pyasn1 package.

Giovanni 
Bye, Giovanni

-------- Messaggio originale --------
Da:Will <notifications@github.com>
Inviato:Wed, 19 Aug 2015 12:34:42 +0200
A:cannatag/ldap3 <ldap3@noreply.github.com>
Cc:Giovanni Cannata <cannatag@gmail.com>
Oggetto:Re: [ldap3] Samba does not honor delete_old_dn=True (#100)

**Comment by cannatag** [comment]
Fixed in 0.9.9
